mater/lib/src/stores/filestore.rs

```diff
@@ -76,18 +84,111 @@ where
     let placeholder_header_v1 = CarV1Header::default();
     position += writer.write_v1_header(&placeholder_header_v1).await?;
 
-    let mut root = None;
+    let mut root: Option<Cid> = None;
     let mut entries = vec![];
+
+    // Collect blocks and track final root
     while let Some(node) = nodes.next().await {
         let (node_cid, node_bytes) = node?;
-        let digest = node_cid.hash().digest().to_owned();
+        let digest = node_cid.hash().digest().to_vec();
+
+        // Index entry for the block
         let entry = IndexEntry::new(digest, (position - car_v1_start) as u64);
         entries.push(entry);
+
+        // Write the block
         position += writer.write_block(&node_cid, &node_bytes).await?;
+        root = Some(node_cid);
+    }
+
+    let index_offset = position;
+    let single_width_index =
+        SingleWidthIndex::new(Sha256::output_size() as u32, entries.len() as u64, entries);
+    let index = Index::MultihashIndexSorted(MultihashIndexSorted::from_single_width(
+        SHA_256_CODE,
+        single_width_index.into(),
+    ));
+    writer.write_index(&index).await?;
+
+    writer.get_inner_mut().rewind().await?;
+    let header = CarV2Header::new(
+        false,
+        car_v1_start as u64,
+        (index_offset - car_v1_start) as u64,
+        index_offset as u64,
+    );
+    writer.write_header(&header).await?;
+
+    let header_v1 = CarV1Header::new(root.into_iter().collect());
+    writer.write_v1_header(&header_v1).await?;
+
+    writer.finish().await?;
+    Ok(header_v1.roots.first().cloned())
+}
+
+/// Similar to [`balanced_import`], this function also creates a “balanced” CARv2 layout.
+// Unlike [`balanced_import`], it handles UnixFS blocks and ensures there is always at
+// least one root (returning an error otherwise).
+// We keep this separate from [`balanced_import`] because:
+// UnixFS Data: This function encodes blocks differently, wrapping them with UnixFS
+//   metadata rather than raw.
+// Guaranteed Root: We return an error if no root is found, whereas [`balanced_import`]
+//   allows for an empty result (`Option<Cid>`).
+// Implementation Clarity: Combining both approaches into a single function would
+//   introduce extra branching and logic. If we find a cleaner abstraction in the future,
+//   we may unify them.
+async fn balanced_import_unixfs<Src, Out>(
+    mut source: Src,
+    mut output: Out,
+    chunk_size: usize,
+    tree_width: usize,
+) -> Result<Cid, Error>
+where
+    Src: AsyncRead + Unpin,
+    Out: AsyncWrite + AsyncSeek + Unpin,
+{
+    let chunker = async_stream::try_stream! {
+        let mut buf = BytesMut::with_capacity(chunk_size);
+        loop {
+            let read_bytes = source.read_buf(&mut buf).await?;
```

Comment:
If we improved this implementation, why it wasn't needed to change the same chunker in the `balanced_import`?